polkadot/node/network/collator-protocol/src/validator_side/collation.rs

```diff
@@ -216,27 +221,97 @@ impl CollationStatus {
 	/// How many collations have been seconded.
 	pub seconded_count: usize,
+	/// What collations were fetched so far for this relay parent.
+	fetched_per_para: BTreeMap<ParaId, usize>,
+	// Claims per `ParaId` for the assigned core at the relay parent. This information is obtained
+	// from `GroupAssignments` which contains either the claim queue (if runtime supports it) for
+	// the core or the `ParaId` of the parachain assigned to the core.
+	claims_per_para: BTreeMap<ParaId, usize>,
+	// Represents the claim queue at the relay parent. The `bool` field indicates if a candidate
+	// was fetched for the `ParaId` at the position in question. In other words - if the claim is
+	// 'satisfied'. If the claim queue is not available `claim_queue_state` will be `None`.
+	claim_queue_state: Option<Vec<(bool, ParaId)>>,
 }
 
 impl Collations {
+	/// `Collations` should work with and without claim queue support. If the claim queue runtime
+	/// api is available `GroupAssignments` the claim queue. If not - group assignments will contain
+	/// just one item (what's scheduled on the core).
+	///
+	/// Some of the logic in `Collations` relies on the claim queue and if it is not available
+	/// fallbacks to another logic. For this reason `Collations` needs to know if claim queue is
+	/// available or not.
+	///
+	/// Once claim queue runtime api is released everywhere this logic won't be needed anymore and
+	/// can be cleaned up.
+	pub(super) fn new(group_assignments: &Vec<ParaId>, has_claim_queue: bool) -> Self {
+		let mut claims_per_para = BTreeMap::new();
+		let mut claim_queue_state = Vec::with_capacity(group_assignments.len());
+
+		for para_id in group_assignments {
+			*claims_per_para.entry(*para_id).or_default() += 1;
+			claim_queue_state.push((false, *para_id));
+		}
+
+		// Not optimal but if the claim queue is not available `group_assignments` will have just
+		// one element. Can be fixed once claim queue api is released everywhere and the fallback
+		// code is cleaned up.
+		let claim_queue_state = if has_claim_queue { Some(claim_queue_state) } else { None };
+
+		Self {
+			status: Default::default(),
+			fetching_from: None,
+			waiting_queue: Default::default(),
+			seconded_count: 0,
+			fetched_per_para: Default::default(),
+			claims_per_para,
+			claim_queue_state,
+		}
+	}
+
 	/// Note a seconded collation for a given para.
 	pub(super) fn note_seconded(&mut self) {
 		self.seconded_count += 1
 	}
 
+	// Note a collation which has been successfully fetched.
+	pub(super) fn note_fetched(&mut self, para_id: ParaId) {
+		// update the number of fetched collations for the para_id
+		*self.fetched_per_para.entry(para_id).or_default() += 1;
+
+		// and the claim queue state
+		if let Some(claim_queue_state) = self.claim_queue_state.as_mut() {
+			for (satisfied, assignment) in claim_queue_state {
+				if *satisfied {
+					continue
+				}
+
+				if assignment == &para_id {
+					*satisfied = true;
+					break
+				}
```

Comment:
before fetching the collation and seconding it, we ask prospective-parachains whether we may second this candidate. prospective-parachains as of now will not allow forks. However, there is work under way to bring that back: https://github.com/paritytech/polkadot-sdk/pull/4937

Reply:
`can_second` cover these checks and prospective-parachains doesn't allow forks at the moment but this will change in the future. 

I think the current implementation doesn't care about forks - as soon as `max_candidate_depth` candidates are accepted 'we are done'.